### PR TITLE
[redux form] FieldArray optional 'props' prop

### LINKED ISSUE #36903: [redux-form] Incorrect typings for FieldArray
If you know how to fix the issue, make a pull request instead.

- [X] I tried using the `@types/xxxx` package and had problems.
- [X] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [ ] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [X] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @m-b-davis @aikoven @LKay @bancek @Aleksion @weswigham

Using a `<FieldArray>` to create custom class-based component in TypeScript 3.5+ fails. Something wrong with props interfaces (`WrappedFieldArrayProps`, `BaseFieldArrayProps`). Here is an example: https://codesandbox.io/s/kind-shadow-us55u

Installed dependencies:
- redux-form@8.2.4
- @types/redux-form@8.1.4

If you are trying to use `BaseFieldArrayProps` with `WrappedFieldArrayProps`, the following error appears:
```
Type 'typeof CustomComponent' is not assignable to type 'ComponentClass<WrappedFieldArrayProps<any> & { name: "any_string"; component: typeof CustomComponent; customProp: string; }, any> | FunctionComponent<WrappedFieldArrayProps<any> & { name: "any_string"; component: typeof CustomComponent; customProp: string; }> | (typeof CustomComponent & ComponentClass<...>) | (typeo...'.
  Type 'typeof CustomComponent' is not assignable to type 'ComponentClass<WrappedFieldArrayProps<any> & { name: "any_string"; component: typeof CustomComponent; customProp: string; }, any>'.
    Type 'CustomComponent' is not assignable to type 'Component<WrappedFieldArrayProps<any> & { name: "any_string"; component: typeof CustomComponent; customProp: string; }, any, any>'.
      Types of property 'props' are incompatible.
        Type 'Readonly<CustomComponentProps> & Readonly<{ children?: ReactNode; }>' is not assignable to type 'Readonly<WrappedFieldArrayProps<any> & { name: "any_string"; component: typeof CustomComponent; customProp: string; }> & Readonly<{ children?: ReactNode; }>'.
          Type 'Readonly<CustomComponentProps> & Readonly<{ children?: ReactNode; }>' is not assignable to type 'Readonly<WrappedFieldArrayProps<any> & { name: "any_string"; component: typeof CustomComponent; customProp: string; }>'.
            Types of property 'name' are incompatible.
              Type 'string' is not assignable to type '"any_string"'.ts(2322)
``` 

You couldn't use only `BaseFieldArrayProps`, because you need `fields` and `meta` props. With only `WrappedFieldArrayProps` you must define `name` and `component` props. I don't understand, how to fix this collision.

### PR DESCRIPTION
**Pull request by habovh**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://redux-form.com/8.3.0/docs/api/fieldarray.md/#-code-props-object-code-optional-

Fixes #36903

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## redux-form ([<kbd>unpkg</kbd>](https://unpkg.com/browse/redux-form@latest/))
was missing the following properties:
1. defaultShouldAsyncValidate
1. defaultShouldValidate
1. defaultShouldError
1. defaultShouldWarn
1. propTypes

<details>
<summary>as well as these 8 other properties...</summary>
<p>fieldInputPropTypes, fieldMetaPropTypes, fieldPropTypes, fieldArrayFieldsPropTypes, fieldArrayMetaPropTypes, fieldArrayPropTypes, formPropTypes, values</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 6baa86d07e4762ca573366fa85e2aab8b01583cb
</p>

File: types/redux-form/lib/FieldArray.d.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ interface _BaseFieldArrayProps<P = {}, FieldValue = any> {
     rerenderOnEveryChange?: boolean | undefined;
 }
 
-export type BaseFieldArrayProps<P = {}, FieldValue = any> = (P | { props: P}) & _BaseFieldArrayProps<P, FieldValue>;
```
Comment (amcasey) [inline types/redux-form/lib/FieldArray.d.ts:13]: I'm pretty unfamiliar with these types, but the current behavior definitely looks intentional.  In some of the tests under the one you added, you can see that two approaches are used - merging properties into the parent or adding them under a new `props` property.

If I'm reading 7c28be5245fa6959a8f587e1410ef12c261b02d8 correctly, it looks like the need to insert an ugly `<{}>` comes from a type inference failure and not from a problem with the types.

Personally, my inclination would be to leave things as they are, though I'm open to feedback, particularly from the owners of these types.

Having said that, if we go ahead with a fix like this, it seems like it should be `P & { props?: P }` to match (what I assume is a workaround in) `Field.d.ts`.

**Review by amcasey — CHANGES_REQUESTED** [review]

**Comment by habovh** [comment]
@amcasey thanks for taking the time to review this PR! Sorry for taking so long to reply though.

I am not that familiar with these types either, but there is definitely something wrong here.

The problem with leaving things as they are is that if you do not pass any other prop to `<FieldArray />`, then TS will complain. It is true however that as soon as any prop is added TS is happy. If an empty object is passed to the `props` prop, TS is equally happy.

If we were to use the solution you're suggesting `P & { props?: P }`, the `<FieldArrayCustom name="field10" />` test would fail. I must admit I do not see which part of `Field.d.ts` you were referring to as a workaround.

File: types/redux-form/lib/FieldArray.d.ts (hunk 1 above)
Comment (amcasey) [reply types/redux-form/lib/FieldArray.d.ts:13]: After so long, I also can't tell what I was talking about in `Field.d.ts`. 😅 

As I said before, I think the current types are correct and this is a workaround for a TS issue (which is probably already known and, given that it's so old, very difficult to fix).  However, I realize that that doesn't help you and your change seems acceptable.

**Review by amcasey — APPROVED** [review]

**Comment by habovh** [comment]
Ready to merge

**Pull request merged**